// sdk-node-apis-efi/nova-sdk/dist/types/pix.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PixCreateImmediateChargeResponseSchema = exports.PixCreateImmediateChargeBodySchema = exports.PixInfoAdicionalSchema = exports.PixLocSchema = exports.PixValorSchema = exports.PixDevedorSchema = exports.PixCalendarioSchema = void 0;
const zod_1 = require("zod");
// Tipos para o método pixCreateImmediateCharge
exports.PixCalendarioSchema = zod_1.z.object({
    expiracao: zod_1.z.number(),
});
exports.PixDevedorSchema = zod_1.z
    .union([
    zod_1.z
        .object({
        cpf: zod_1.z.string().describe('CPF do devedor (somente números)'),
        nome: zod_1.z.string().describe('Nome do devedor'),
    })
        .strict(),
    zod_1.z
        .object({
        cnpj: zod_1.z.string().describe('CNPJ do devedor (somente números)'),
        nome: zod_1.z.string().describe('Nome do devedor'),
    })
        .strict(),
])
    .optional();
exports.PixValorSchema = zod_1.z.object({
    original: zod_1.z.string(),
});
exports.PixLocSchema = zod_1.z.object({
    id: zod_1.z.number(),
});
exports.PixInfoAdicionalSchema = zod_1.z.object({
    nome: zod_1.z.string(),
    valor: zod_1.z.string(),
});
exports.PixCreateImmediateChargeBodySchema = zod_1.z.object({
    calendario: exports.PixCalendarioSchema,
    devedor: exports.PixDevedorSchema,
    valor: exports.PixValorSchema,
    chave: zod_1.z.string(),
    solicitacaoPagador: zod_1.z.string().optional(),
    loc: exports.PixLocSchema.optional(),
    infoAdicionais: zod_1.z.array(exports.PixInfoAdicionalSchema).optional(),
});
exports.PixCreateImmediateChargeResponseSchema = zod_1.z.object({
    calendario: zod_1.z.object({
        criacao: zod_1.z.string(),
        expiracao: zod_1.z.number(),
    }),
    txid: zod_1.z.string(),
    revisao: zod_1.z.number(),
    loc: zod_1.z.object({
        id: zod_1.z.number(),
        location: zod_1.z.string(),
        tipoCob: zod_1.z.string(),
    }),
    location: zod_1.z.string(),
    status: zod_1.z.string(),
    devedor: exports.PixDevedorSchema,
    valor: exports.PixValorSchema,
    chave: zod_1.z.string(),
    solicitacaoPagador: zod_1.z.string().optional(),
    pixCopiaECola: zod_1.z.string(),
});
